### PR TITLE
extend test_change_pageserver for failure case, rework changing pageserver

### PR DESCRIPTION
**Pull request by jcsp**
Reproducer for https://github.com/neondatabase/neon/issues/5692

The test change in this PR intentionally fails, to demonstrate the issue.

### CODE REVIEW

File: pgxn/neon/libpagestore.c [hunk 1]
```diff
@@ -61,23 +64,63 @@ int			flush_every_n_requests = 8;
 int			n_reconnect_attempts = 0;
 int			max_reconnect_attempts = 60;
 
+#define MAX_PAGESERVER_CONNSTRING_SIZE 256
+
+typedef struct
+{
+    LWLock lock;
```
Comment (knizhnik) [inline pgxn/neon/libpagestore.c:71]: I do not see place this `lock` is initialized.
Please look at file_cache.c for example of `lfc_lock` registration and initialization.

Comment (save-buffer) [reply pgxn/neon/libpagestore.c]: Thank you, I wasn't sure how to initialize it. I believe I fixed it

File: pgxn/neon/libpagestore.c [hunk 2]
```diff
@@ -405,20 +461,78 @@ check_neon_id(char **newval, void **extra, GucSource source)
...
+PagestoreShmemInit(void)
+{
+    bool found;
+    LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+    pagestore_shared = ShmemInitStruct("libpagestore shared state",
+                                       PagestoreShmemSize(),
+                                       &found);
+    if(!found)
+    {
+        memset(pagestore_shared, 0, PagestoreShmemSize());
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pgxn/neon/libpagestore.c]: No longer use memset to initialize

File: pgxn/neon/libpagestore.c [hunk 3]
```diff
@@ -61,23 +64,63 @@ int			flush_every_n_requests = 8;
...
-	if (prev_signal_handler)
-	{
-        	prev_signal_handler(postgres_signal_arg);
-	}
-	neon_log(LOG, "Received SIGHUP, disconnecting pageserver. New pageserver connstring is %s", page_server_connstring);
-	pageserver_disconnect();
+    if(!pagestore_shared)
+        return;
+    LWLockAcquire(&pagestore_shared->lock, LW_SHARED);
+    strlcpy(local_pageserver_connstring, pagestore_shared->pageserver_connstring, MAX_PAGESERVER_CONNSTRING_SIZE);
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pgxn/neon/libpagestore.c]: I don't know why this is going off, given that I am already using `strlcpy` as it suggests, but I changed it to sizeof(local_pageserver_connstring).

**Review by knizhnik — APPROVED** [review]

**Pull request merged**